Quickstart installation of akka operator on GCP using Pulumi templating (#720)
akka/akka-platform-guide: docs-source/docs/modules/deployment/pages/gcp-install-quickstart.adoc

= Installation on Google Kubernetes Engine (GKE) Quick Start
:page-toclevels: 3

include::partial$include.adoc[]

> NOTE: To install Akka Cloud Platform on Google Kubernetes Engine (GKE), you must have a Google account and a subscription to the {gcp-marketplace}[Akka Cloud Platform {tab-icon}, window="tab"].

Reviewer: Maybe add a link how to get the subscription?

Author: @RayRoestenburg What link should we add here ?

== Expected Time to Install

A new installation will take approximately  _20 minutes_ and the following components will be provisioned:

1. Create a Google GKE cluster
2. Install the Akka Cloud Platform Helm chart
3. Install a PostgreSQL Database Instance

:sectnums:

== Verify Prerequisites

Before installation, you must subscribe to the {gcp-marketplace}[Akka Cloud Platform {tab-icon}, window="tab"] on GCP Marketplace by clicking the Subscribe button on the product page and accepting the terms and conditions.

To install and use the Akka Cloud Platform, you must have the following tools installed. We recommend using the latest versions of these tools. If you do not have them yet or need to update, we provide links to installation instructions:

* The https://git-scm.com/book/en/v2/Getting-Started-Installing-Git[`git` {tab-icon}, window="tab"] command-line tool to clone the Pulumi Playbook.

* The https://docs.npmjs.com/downloading-and-installing-node-js-and-npm[`npm` {tab-icon}, window="tab"] command-line tool to download and setup the NPM dependencies of the Pulumi Playbook.

> * The https://www.pulumi.com/[Pulumi {tab-icon}, window="tab"] cloud engineering (provisioning) tool to run the Playbook.

Reviewer: ```suggestion
* The https://www.pulumi.com/[pulumi {tab-icon}, window="tab"] cloud engineering (provisioning) tool to run the Playbook.
```
(since you also did `git` and `npm`

Author: done ..

In addition, in order to operate the cluster, you need to setup:

> * The Kubernetes command-line tool, `kubectl`, allows you to run commands against Kubernetes clusters. Follow the instructions in the https://kubernetes.io/docs/tasks/tools/#kubectl[Kubernetes documentation {tab-icon}, window="tab"]  to install `kubectl`.

Reviewer: ```suggestion
* The Kubernetes command-line tool, `kubectl`, which allows you to run commands against Kubernetes clusters. Follow the instructions in the https://kubernetes.io/docs/tasks/tools/#kubectl[Kubernetes documentation {tab-icon}, window="tab"]  to install `kubectl`.
```

Author: done ..

Reviewer 2: Um,  no. Aiding the 'which' requires that you add a subordinate clause

Reviewer 2: @marcospereira do you mean we need to add the entire paragraph to the AWS section?

Reviewer 3: @thothmoses, the paragraph is already there, but if we need to add the `which` here, it makes sense to add it there too. I'm not an English native speaker, so I will defer to you if this is necessary or not.

== Setup your Google Cloud Platform (GCP)

The https://www.pulumi.com/docs/intro/cloud-providers/gcp[Pulumi Google Cloud Platform Provider {tab-icon}, window="tab"] needs to be configured with Google credentials before it can be used to create resources. Ensure the following steps are done before going ahead with creation of the cluster:

* Install the https://cloud.google.com/sdk/install[Google Cloud SDK {tab-icon}, window="tab"] and then https://cloud.google.com/sdk/docs/authorizing#authorizing_with_a_user_account[authorize access with a user account {tab-icon}, window="tab].

* Configure `gcloud` to interact with your Google Cloud project using the `config` command:

[source, shell script]
----
gcloud config set project <YOUR_GCP_PROJECT_ID>
----

* Setup default application credentials to interact with your Google Cloud resources. Pulumi requires this.

[source, shell script]
----
gcloud auth application-default login
----

> * Some optional settings can be done for ready defaults. The following environment variables can be set to provide optional configurations to Pulumi Google Cloud Platform provider:

Reviewer: What do you mean with `ready defaults`?

Author: The next sentence explains what I mean by __ready defaults__. Setting those environment variables provide optional configurations to Pulumi GCP provider.

  . GOOGLE_PROJECT - The default project for new resources, if one is not specified when creating a resource
  . GOOGLE_REGION - The default region for new resources, if one is not specified when creating a resource
  . GOOGLE_ZONE - The default zone for new resources, if one is not specified when creating a resource.

Alternatively these can also be set through `pulumi config` to make them defaults for newly created resources:

[source, shell script]

> ----

Reviewer: Are commands further on going to depend on these? since it is optional I guess not? I'm not sure if or when the reader needs these?

Author: Instead of setting environment variables, the user can use `pulumi config set` to set these values. The playbook will use these config.

To begin, download and install the https://www.pulumi.com/docs/get-started/install/[Pulumi {tab-icon}, window="tab"] cloud engineering (provisioning) tool for your platform.

Download (clone) the https://github.com/lightbend/akka-cloud-platform-deploy[Akka Cloud Platform Deployment Pulumi Playbook {tab-icon}, window="tab"] with https://git-scm.com/book/en/v2/Getting-Started-Installing-Git[`git` {tab-icon}, window="tab"].

[source,shell script]
----
git clone https://github.com/lightbend/akka-cloud-platform-deploy
----

Navigate to the base directory of the cloned repository in a terminal.

[source,shell script]
----
cd akka-cloud-platform-deploy
----

The same repository contains the scripts for AWS and GCP deployment. Navigate to the gcp folder.

[source,shell script]
----
cd gcp
----

Install `npm` dependencies:

[source,shell script]
----
npm install
----

Next you will initialize a https://www.pulumi.com/docs/intro/concepts/stack/[Pulumi Stack {tab-icon}, window="tab"].
A stack is an instance of the deployed Pulumi playbook and its associated state.
For example, you might use different stacks to represent different environments such as `dev`, `testing`, and `staging`.

> By default state is managed using Pulumi's state management service (a remote service), but you can choose a https://www.pulumi.com/docs/intro/concepts/state/#backends[self-managed state backend {tab-icon}, window="tab"] if you like.

Reviewer: ```suggestion
State is managed by default using Pulumi's state management service (a remote service), but you can choose a https://www.pulumi.com/docs/intro/concepts/state/#backends[self-managed state backend {tab-icon}, window="tab"] if you like.
```

Author: done ..

Using Pulumi's service is the fastest way to get started and provides several other benefits such as sharing state with others, view progress of a deployment, and view output variables and the resulting cloud resources that were created.
To learn more about https://www.pulumi.com/docs/intro/concepts/stack/[Pulumi Stack {tab-icon}, window="tab"] and https://www.pulumi.com/docs/intro/concepts/state/[Pulumi State Management {tab-icon}, window="tab"] see the Pulumi documentation.

Initialize a new Pulumi stack.

[source,shell script]
----

> pulumi stack init

Reviewer: The stack-name is missing in this command

----

Choose Google Project Id, zone and region to use:

[source,shell script]
----
pulumi config set gcp:project <google project id>

> pulumi config set gcp:zone europe-west1-b

Reviewer: Are zone and region necessary? (if you leave it out it does not work?)

Author: If they r not specified the values will be picked up from environment variables

pulumi config set gcp:region europe-west1
----

> Setup the license key file path. This is required to be set in config before you run the playbook. This sets up the license key file path which will be passed to the operator deployment process. If you are deploying from local environment you need to download the license file once from GCP Marketplace as per instructions in the xref:gcp-install.adoc[installation guide].

Reviewer: ```suggestion
Setup the license key file path. This is required to be set before you run the playbook. This sets up the license key file path which will be passed to the operator deployment process. If you are deploying from local environment you need to download the license file once from GCP Marketplace as per instructions in the xref:gcp-install.adoc[installation guide].
```

Reviewer: Its not really clear how to get the license key. Can you inline the instruction?

Reviewer: (Also it would be great to add link to marketplace page: https://console.cloud.google.com/marketplace/kubernetes/config/lightbend-public/akka-cloud-platform)

Author: done ..

[source,shell script]
----

> pulumi config set akka-cloud-platform-gcp-deploy:license-file-path "/<path>/<to>/license.yaml"

Reviewer: Are the quotes required?

Author: No, unless you have some special characters in the path string. Hence just to be safe ..

...
----

When the stack is deployed successfully you will have a list of outputs presented. Example outputs for a stack named `dev`:

[source,shell script]
----
Outputs:
    clusterName                       : "projects/bubbly-observer-178213/locations/europe-west1-b/clusters/acp-dev-gke-111241b"
    kubeconfig                        : "[secret]"
    operatorNamespace                 : "lightbend"
    postgreSQLEndpoint                : "10.10.0.3"
    postgreSQLInstanceConnectionName  : "bubbly-observer-178213:europe-west1:instance-b787e55"
    postgreSQLInstanceName            : "instance-b787e55"

Resources:
    + 23 created

Duration: 22m39s
----

These output variables are required to access your cluster and associated resources.
You can reference variables manually, or access their values at the command line with `pulumi stack output <OUTPUT KEY>`.
Later in this documentation page you can find the full list of xref:deployment:gcp-install-quickstart.adoc#output-variables[output variables].

For example, to return the Kubernetes namespace that the Akka Cloud Platform Operator is installed in:

[source, shell script]
----
$ pulumi stack output operatorNamespace
lightbend
----

> NOTE: Pulumi has a bug in tracking ready status for containiners with sidecars. In GCP implementation of akka platform operator, ubbagent (the billing agent)

Reviewer: ```suggestion
NOTE: Pulumi has a bug in tracking ready status for containers with sidecars. In GCP implementation of akka platform operator, ubbagent (the billing agent)
```

Author: done

----

To destroy the cluster and all its dependencies:

[source,shell script]
----
pulumi destroy
----

=== Connect to the GKE Cluster

To setup your local `kubectl` configuration you can obtain the Pulumi `kubeconfig` output by exporting it to a file and defining the `KUBECONFIG` environment variable:

[source,shell script]
----
pulumi stack output kubeconfig --show-secrets > kubeconfig.yml
export KUBECONFIG=./kubeconfig.yml
----

[#connect-cloudsql-database]
=== Connect to the Cloud SQL database

To open a Postgres interactive shell with the Cloud SQL database instance:

First you may want to set up a password for the Postgres user that you will be using. Here's how to do that for the default user `postgres`

[source,shell script]
----
gcloud sql users set-password postgres --instance=$(pulumi stack output postgreSQLInstanceName) --prompt-for-password
----

Once the password is set, use `kubectl run` to set up an interactive `psql` session on a temporary pod:

[source,shell script]
----
kubectl run -i --tty rds-mgmt --image=postgres --restart=Never --rm \
  --env=PGPASSWORD=<password> -- \
  psql -h $(pulumi stack output postgreSQLEndpoint) -U postgres
----

> 

Reviewer: maybe add output variables and configuration of playbook sections like in the AWS install quickstart?

Author: done

= Installation on Google Kubernetes Engine (GKE) Quick Start
:page-toclevels: 3

include::partial$include.adoc[]

NOTE: To install Akka Cloud Platform on Google Kubernetes Engine (GKE), you must have a Google account and a subscription to the {gcp-marketplace}[Akka Cloud Platform {tab-icon}, window="tab"].

== Expected Time to Install

A new installation will take approximately  _20 minutes_ and the following components will be provisioned:

> 1. Create a Google GKE cluster
> 2. Install the Akka Cloud Platform Helm chart
> 3. Install a PostgreSQL Database Instance

Reviewer 2: ```suggestion
1. Create a Google GKE cluster.
2. Install the Akka Cloud Platform Helm chart.
3. Install a PostgreSQL Database Instance.
```

Author: 👍

:sectnums:

== Verify Prerequisites

Before installation, you must subscribe to the {gcp-marketplace}[Akka Cloud Platform {tab-icon}, window="tab"] on GCP Marketplace by clicking the Subscribe button on the product page and accepting the terms and conditions.

To install and use the Akka Cloud Platform, you must have the following tools installed. We recommend using the latest versions of these tools. If you do not have them yet or need to update, we provide links to installation instructions:

* The https://git-scm.com/book/en/v2/Getting-Started-Installing-Git[`git` {tab-icon}, window="tab"] command-line tool to clone the Pulumi Playbook.

* The https://docs.npmjs.com/downloading-and-installing-node-js-and-npm[`npm` {tab-icon}, window="tab"] command-line tool to download and setup the NPM dependencies of the Pulumi Playbook.

* The https://www.pulumi.com/[pulumi {tab-icon}, window="tab"] cloud engineering (provisioning) tool to run the Playbook.

In addition, in order to operate the cluster, you need to setup:

> * The Kubernetes command-line tool, `kubectl`, which allows you to run commands against Kubernetes clusters. Follow the instructions in the https://kubernetes.io/docs/tasks/tools/#kubectl[Kubernetes documentation {tab-icon}, window="tab"]  to install `kubectl`.

Reviewer 2: ```suggestion
* The Kubernetes command-line tool, `kubectl`, allows you to run commands against Kubernetes clusters. Follow the instructions in the https://kubernetes.io/docs/tasks/tools/#kubectl[Kubernetes documentation {tab-icon}, window="tab"]  to install `kubectl`.
```

Author: 👍

== Setup your Google Cloud Platform (GCP)

> The https://www.pulumi.com/docs/intro/cloud-providers/gcp[Pulumi Google Cloud Platform Provider {tab-icon}, window="tab"] needs to be configured with Google credentials before it can be used to create resources. Ensure the following steps are done before going ahead with creation of the cluster:

Reviewer 2: ```suggestion
The https://www.pulumi.com/docs/intro/cloud-providers/gcp[Pulumi Google Cloud Platform Provider {tab-icon}, window="tab"] needs to be configured with Google credentials before it can be used to create resources. Ensure the following steps are complete before proceding with the creation of the cluster:
```

Author: 👍

* Install the https://cloud.google.com/sdk/install[Google Cloud SDK {tab-icon}, window="tab"] and then https://cloud.google.com/sdk/docs/authorizing#authorizing_with_a_user_account[authorize access with a user account {tab-icon}, window="tab].

* Configure `gcloud` to interact with your Google Cloud project using the `config` command:

[source, shell script]
----
gcloud config set project <YOUR_GCP_PROJECT_ID>
----

> * Setup default application credentials to interact with your Google Cloud resources. Pulumi requires this.

Reviewer 2: ```suggestion
* Pulumi requires that you set up default application credentials in order to interact with your Google Cloud resources. 
```

Author: 👍

[source, shell script]
----
gcloud auth application-default login
----

> * Some optional settings can be done to provide ready defaults. The following environment variables can be set to provide optional configurations to Pulumi Google Cloud Platform provider:

Reviewer 2: ```suggestion
* Some optional settings can be set up to provide ready defaults. The following environment variables can be set to provide optional configurations to the  Pulumi Google Cloud Platform provider:
```

  . GOOGLE_PROJECT - The default project for new resources, if one is not specified when creating a resource
  . GOOGLE_REGION - The default region for new resources, if one is not specified when creating a resource
  . GOOGLE_ZONE - The default zone for new resources, if one is not specified when creating a resource.

Alternatively these can also be set through `pulumi config` to make them defaults for newly created resources:

[source, shell script]
----
pulumi config set gcp:project <your-project-here>
pulumi config set gcp:region <your-region-here>
pulumi config set gcp:zone <your-zone-here>
----

== Quick Start with Pulumi Playbook

Pulumi is a cloud provisioning tool used to set up cloud infrastructure and Kubernetes cluster resources.
The fastest way to get started with Akka Cloud Platform on GCP is to run the Akka Cloud Platform Deployment Pulumi playbook.

> 

Reviewer 2: I would inject a description of what you mean by playbook here. Simply because that is an overloaded term in the industry.

Author: 👍

To begin, download and install the https://www.pulumi.com/docs/get-started/install/[Pulumi {tab-icon}, window="tab"] cloud engineering (provisioning) tool for your platform.

Download (clone) the https://github.com/lightbend/akka-cloud-platform-deploy[Akka Cloud Platform Deployment Pulumi Playbook {tab-icon}, window="tab"] with https://git-scm.com/book/en/v2/Getting-Started-Installing-Git[`git` {tab-icon}, window="tab"].

[source,shell script]
----
git clone https://github.com/lightbend/akka-cloud-platform-deploy
----

Navigate to the base directory of the cloned repository in a terminal.

[source,shell script]
----
cd akka-cloud-platform-deploy
----

The same repository contains the scripts for AWS and GCP deployment. Navigate to the gcp folder.

[source,shell script]
----
cd gcp
----

Install `npm` dependencies:

[source,shell script]
----
npm install

> ----

Reviewer 2: Enumerate the steps - it will make it easier to follow this section

Author: 👍

NOTE: The setting of project id and zone are mandatory for a GCP installation. In case these are not set, the installation will be aborted with appropriate error messages.

Setup the license key file path. This is required to be set before you run the playbook. This sets up the license key file path which will be passed to the operator deployment process. If you are deploying from local environment you need to download the license file once from GCP Marketplace as per instructions in the xref:gcp-install.adoc[installation guide]. In the {gcp-marketplace}[Akka Cloud Platform {tab-icon}, window="tab"] for GCP page, click on __Configure__ and then go to the tab __Deploy via command line__ and click on __Generate license key__. A `license.yaml` will be generated on your local system.

[source,shell script]
----
pulumi config set akka-cloud-platform-gcp-deploy:license-file-path "/<path>/<to>/license.yaml"
----

Setup the version of the operator.

[source,shell script,subs="attributes+"]
----
pulumi config set akka-cloud-platform-gcp-deploy:operator-version {akka-cloud-platform-version}
----

> NOTE: If you forget to set these up, the playbook will give you a gentle reminder to set the values appropriately.

Reviewer 2: 'these' has an unclear antecedent.

Author: Made the context explicit

 +   pulumi:pulumi:Stack                                                  akka-cloud-platform-gcp-deploy-<STACK NAME>                                                 created

...
----

When the stack is deployed successfully you will have a list of outputs presented. Example outputs for a stack named `dev`:

[source,shell script]
----
Outputs:
    clusterName                       : "projects/bubbly-observer-178213/locations/europe-west1-b/clusters/acp-dev-gke-111241b"
    kubeconfig                        : "[secret]"
    operatorNamespace                 : "lightbend"
    postgreSQLEndpoint                : "10.10.0.3"
    postgreSQLInstanceConnectionName  : "bubbly-observer-178213:europe-west1:instance-b787e55"
    postgreSQLInstanceName            : "instance-b787e55"

Resources:
    + 23 created

Duration: 22m39s
----

These output variables are required to access your cluster and associated resources.
You can reference variables manually, or access their values at the command line with `pulumi stack output <OUTPUT KEY>`.

> Later in this documentation page you can find the full list of xref:deployment:gcp-install-quickstart.adoc#output-variables[output variables].

Reviewer 2: ```suggestion
Later on this page you can find the full list of xref:deployment:gcp-install-quickstart.adoc#output-variables[output variables].
```

For example, to return the Kubernetes namespace that the Akka Cloud Platform Operator is installed in:

[source, shell script]
----
$ pulumi stack output operatorNamespace
lightbend
----

> NOTE: Pulumi has a bug in tracking ready status for containers with sidecars. In GCP implementation of akka platform operator, ubbagent (the billing agent)

Reviewer 2: ```suggestion
NOTE: Pulumi has an issue with tracking ready status for containers with sidecars. In GCP implementation of the akka platform operator, ubbagent (the billing agent)
```

Author: 👍

is implemented as a sidecar. Hence you may see the following failure in `pulumi up` where the update of the stack fails for akka-operator:

[source,shell script]
----
Diagnostics:
  kubernetes:apps/v1:Deployment (lightbend/akka-operator):
    error: 4 errors occurred:
    	* resource lightbend/akka-operator was successfully created, but the Kubernetes API server reported that it failed to fully initialize or become live: 'akka-operator' timed out waiting to be Ready
    	* [MinimumReplicasUnavailable] Deployment does not have minimum availability.
    	* Minimum number of live Pods was not attained
    	* [Pod lightbend/akka-operator-9b6f567cc-psqhh]: containers with unready status: [akka-operator ubbagent]
 
  pulumi:pulumi:Stack (akka-cloud-platform-gcp-deploy-dev):
    error: update failed
----

Just run `pulumi up` again which is basically a noop but the stack gets updated fully.

> NOTE: Another documented bug in Pulumi is a problem in concurrency control where an operation fails with the following message. This is also documented in https://www.pulumi.com/docs/troubleshooting/#conflict[pulumi troubleshooting section {tab-icon}, window="tab"]

Reviewer 2: ```suggestion
NOTE: Another documented issue in Pulumi is a problem in concurrency control. In this case, an operation fails with the following message. This is also documented in https://www.pulumi.com/docs/troubleshooting/#conflict[pulumi troubleshooting section {tab-icon}, window="tab"]
```

Author: 👍

[source,shell script]
----
gcp:sql:DatabaseInstance (pg12instance):
  error: 1 error occurred:
  	* updating urn:pulumi:dev::akka-cloud-platform-gcp-deploy::gcp:sql/databaseInstance:DatabaseInstance::pg12instance: 1 error occurred:
  	* Error, failed to update instance settings for : googleapi: Error 409: Operation failed because another operation was already in progress., operationInProgress
 
 pulumi:pulumi:Stack (akka-cloud-platform-gcp-deploy-dev):
  error: update failed
----

> Doing a `pulumi cancel` will cancel the running update. However in some cases the update might have gone ok despite the above message. In that case `pulumi cancel` gives the following message __The Update has already completed__.

Reviewer 2: ```suggestion
Doing a `pulumi cancel` will cancel the running update. However in some cases, the update might have gone 'OK', despite the message mentioned above. In that case, `pulumi cancel` gives the following message __The Update has already completed__.
```

Author: 👍

[source,shell script]
----
pulumi cancel

This will irreversibly cancel the currently running update for 'dev'!
Please confirm that this is what you'd like to do by typing ("dev"): dev
error: [409] Conflict: The Update has already completed
----

To destroy the cluster and all its dependencies:

[source,shell script]
----
pulumi destroy
----

=== Connect to the GKE Cluster

> To setup your local `kubectl` configuration you can obtain the Pulumi `kubeconfig` output by exporting it to a file and defining the `KUBECONFIG` environment variable:

Reviewer 2: ```suggestion
To set up your local `kubectl` configuration, you can obtain the Pulumi `kubeconfig` output by exporting it to a file, and defining the `KUBECONFIG` environment variable:
```

Author: 👍

[source,shell script]
----
pulumi stack output kubeconfig --show-secrets > kubeconfig.yml
export KUBECONFIG=./kubeconfig.yml
----

[#connect-cloudsql-database]
=== Connect to the Cloud SQL database

To open a Postgres interactive shell with the Cloud SQL database instance:

> First you may want to set up a password for the Postgres user that you will be using. Here's how to do that for the default user `postgres`

Reviewer 2: ```suggestion
First, you may want to set up a password for the Postgres user that you will be using. Here's how to do that for the default user `postgres`
```

Author: 👍

[source,shell script]
----
gcloud sql users set-password postgres --instance=$(pulumi stack output postgreSQLInstanceName) --prompt-for-password
----

Once the password is set, use `kubectl run` to set up an interactive `psql` session on a temporary pod:

[source,shell script]
----
kubectl run -i --tty rds-mgmt --image=postgres --restart=Never --rm \
  --env=PGPASSWORD=<password> -- \
  psql -h $(pulumi stack output postgreSQLEndpoint) -U postgres
----

[#output-variables]
== Output Variables

> To print all output variables for a deployed stack you can run the following command:

Reviewer 2: ```suggestion
To print all output variables for a deployed stack, you can run the following command:
```

Author: 👍

[source,shell script]
----
pulumi stack output
----

> To output the value of only one variable use the same command and include the variable. For example, to return the clusterName.

Reviewer 2: ```suggestion
To output, the value of only one variable, use the same command and include the variable. For example, to return the clusterName.
```

Author: 👍

= Installation on Google Kubernetes Engine (GKE) Quick Start
:page-toclevels: 3

include::partial$include.adoc[]

NOTE: To install Akka Cloud Platform on Google Kubernetes Engine (GKE), you must have a Google account and a subscription to the {gcp-marketplace}[Akka Cloud Platform {tab-icon}, window="tab"].

== Expected Time to Install

> A new installation will take approximately  _20 minutes_ and the following components will be provisioned:

Author: @marcospereira It installs less stuff .. no Kafka :-)

1. Create a Google GKE cluster.
2. Install the Akka Cloud Platform Helm chart.
3. Install a PostgreSQL Database Instance.

:sectnums:

> == Verify Prerequisites

Author: 👍

> Before installation, you must subscribe to the {gcp-marketplace}[Akka Cloud Platform {tab-icon}, window="tab"] on GCP Marketplace by clicking the Subscribe button on the product page and accepting the terms and conditions.

Author: 👍

* The https://git-scm.com/book/en/v2/Getting-Started-Installing-Git[`git` {tab-icon}, window="tab"] command-line tool to clone the Pulumi Playbook.

* The https://docs.npmjs.com/downloading-and-installing-node-js-and-npm[`npm` {tab-icon}, window="tab"] command-line tool to download and setup the NPM dependencies of the Pulumi Playbook.

* The https://www.pulumi.com/[pulumi {tab-icon}, window="tab"] cloud engineering (provisioning) tool to run the Playbook.

In addition, in order to operate the cluster, you need to setup:

* The Kubernetes command-line tool, `kubectl`, allows you to run commands against Kubernetes clusters. Follow the instructions in the https://kubernetes.io/docs/tasks/tools/#kubectl[Kubernetes documentation {tab-icon}, window="tab"]  to install `kubectl`.

== Setup your Google Cloud Platform (GCP)

The https://www.pulumi.com/docs/intro/cloud-providers/gcp[Pulumi Google Cloud Platform Provider {tab-icon}, window="tab"] needs to be configured with Google credentials before it can be used to create resources. Ensure the following steps are complete before proceeding with the creation of the cluster:

> * Install the https://cloud.google.com/sdk/install[Google Cloud SDK {tab-icon}, window="tab"] and then https://cloud.google.com/sdk/docs/authorizing#authorizing_with_a_user_account[authorize access with a user account {tab-icon}, window="tab].

Author: Added this as a prerequisite in the prerequisites section.

* Configure `gcloud` to interact with your Google Cloud project using the `config` command:

[source, shell script]
----
gcloud config set project <YOUR_GCP_PROJECT_ID>
----

* Pulumi requires that you setup default application credentials in order to interact with your Google Cloud resources. Pulumi requires this.

[source, shell script]
----
gcloud auth application-default login
----

* Some optional settings can be set up to provide ready defaults. The following environment variables can be set to provide optional configurations to the Pulumi Google Cloud Platform provider:
  . GOOGLE_PROJECT - The default project for new resources, if one is not specified when creating a resource
  . GOOGLE_REGION - The default region for new resources, if one is not specified when creating a resource
  . GOOGLE_ZONE - The default zone for new resources, if one is not specified when creating a resource.

Alternatively these can also be set through `pulumi config` to make them defaults for newly created resources:

[source, shell script]
----
pulumi config set gcp:project <your-project-here>
pulumi config set gcp:region <your-region-here>
pulumi config set gcp:zone <your-zone-here>
----

> == Quick Start with Pulumi Playbook

Author: I followed the AWS doc template here.

[source,shell script]
----
pulumi config set akka-cloud-platform-gcp-deploy:license-file-path "/<path>/<to>/license.yaml"
----

* Setup the version of the operator.

[source,shell script,subs="attributes+"]
----
pulumi config set akka-cloud-platform-gcp-deploy:operator-version {akka-cloud-platform-version}
----

NOTE: If you forget to set the license key file path and operator version up, the playbook will give you a gentle reminder to set the values appropriately.

When you run a `pulumi up` (deploy) all the resources defined in the playbook will be provisioned to your target environment.

> In almost all cases resources will start with a prefix `acp-<STACK NAME>`, where `acp` stands for "Akka Cloud Platform" and `<STACK NAME>` is the name you chose for the Pulumi stack.

Author: 👍

Most resource names will end with a randomly generated alphanumeric string.

When you run `pulumi up` you will be prompted with a list of resources that the playbook will create.
Select `yes` to deploy the stack to GCP.

> Preview and provision the playbook.

Author: 👍

[source,shell script]
----
pulumi up
----

While the cluster is provisioning you will be informed, in the terminal, about which resources are in the process of being created or have been created.

> If you are using the Pulumi service for the state then the `pulumi up` command will include a link to "View Live" the progress of the deployment.

Author: 👍

The `pulumi up` command will exit when all resources are fully deployed and online.

If a failure occurs, the command will exit early and inform you of the reason.

> When this occurs and you can correct the issue yourself then rerun `pulumi up` to resume updating where you left off.

Author: 👍

If the error can't be resolved, you can use the `pulumi destroy` command to roll back any partially deployed resources.

If the deployment is successful, all of the pertinent details of the cluster will be output to the console (except for passwords).

> After executing `pulumi up` you will see something like the following:

Author: 👍

[source,shell script]
----
$ pulumi up
Previewing update (dev)

View Live: https://app.pulumi.com/<STACK NAME>/akka-cloud-platform-gcp-deploy/<STACK NAME>/previews/7d82583e-17db-4d80-be58-49d266029998

     Type                                                                 Name                                                                                        Plan

>  +   pulumi:pulumi:Stack                                                  akka-cloud-platform-gcp-deploy-<STACK NAME>                                                 create

Author: @marcospereira Resources' names like cluster name (`acp-dev-gke`) and service account (`acp-dev-sa`) starts with `acp-`, the Stack name is different. Do I need to make any change here ?

>  +   pulumi:pulumi:Stack                                                  akka-cloud-platform-gcp-deploy-<STACK NAME>                                                 create

Reviewer 3: I don't think so. I was mixing different things. Sorry.

[source,shell script]
----
Outputs:
    clusterName                       : "projects/bubbly-observer-178213/locations/europe-west1-b/clusters/acp-dev-gke-111241b"
    kubeconfig                        : "[secret]"
    network                           : "projects/bubbly-observer-178213/global/networks/default"
    operatorNamespace                 : "lightbend"
    postgreSQLEndpoint                : "10.10.0.3"
    postgreSQLInstanceConnectionName  : "bubbly-observer-178213:europe-west1:instance-b787e55"
    postgreSQLInstanceName            : "instance-b787e55"

Resources:
    + 23 created

Duration: 22m39s
----

These output variables are required to access your cluster and associated resources.
You can reference variables manually, or access their values at the command line with `pulumi stack output <OUTPUT KEY>`.
Later on this page you can find the full list of xref:deployment:gcp-install-quickstart.adoc#output-variables[output variables].

For example, to return the Kubernetes namespace that the Akka Cloud Platform Operator is installed in:

[source, shell script]
----
$ pulumi stack output operatorNamespace
lightbend
----

> NOTE: Pulumi has an issue with tracking ready status for containers with sidecars. In GCP implementation of the akka platform operator, ubbagent (the billing agent)

Author: 👍

is implemented as a sidecar. Hence you may see the following failure in `pulumi up` where the update of the stack fails for akka-operator:

[source,shell script]
----
Diagnostics:
  kubernetes:apps/v1:Deployment (lightbend/akka-operator):
    error: 4 errors occurred:
    	* resource lightbend/akka-operator was successfully created, but the Kubernetes API server reported that it failed to fully initialize or become live: 'akka-operator' timed out waiting to be Ready
    	* [MinimumReplicasUnavailable] Deployment does not have minimum availability.
    	* Minimum number of live Pods was not attained
    	* [Pod lightbend/akka-operator-9b6f567cc-psqhh]: containers with unready status: [akka-operator ubbagent]
 
  pulumi:pulumi:Stack (akka-cloud-platform-gcp-deploy-dev):
    error: update failed
----

> Just run `pulumi up` again which is basically a noop but the stack gets updated fully.

Author: 👍

NOTE: Another documented issue in Pulumi is a problem in concurrency control. In this case, an operation fails with the following message. This is also documented in https://www.pulumi.com/docs/troubleshooting/#conflict[pulumi troubleshooting section {tab-icon}, window="tab"]

[source,shell script]
----
gcp:sql:DatabaseInstance (pg12instance):
  error: 1 error occurred:
  	* updating urn:pulumi:dev::akka-cloud-platform-gcp-deploy::gcp:sql/databaseInstance:DatabaseInstance::pg12instance: 1 error occurred:
  	* Error, failed to update instance settings for : googleapi: Error 409: Operation failed because another operation was already in progress., operationInProgress
 
 pulumi:pulumi:Stack (akka-cloud-platform-gcp-deploy-dev):
  error: update failed
----

> Doing a `pulumi cancel` will cancel the running update. However in some cases the update might have gone __'OK'__ despite the message mentioned above. In that case, `pulumi cancel` gives the following message __The Update has already completed__.

Author: 👍

----

The following output variables are available:

GCP GKE

* `clusterName` - Cluster name
* `kubeconfig` - Complete contents of a `KUBECONFIG` configuration for the provisioned cluster as a secret. Need to use `pulumi stack output kubeconfig --show-secrets` to get the list.
* `network` - The name or self_link of the Google Compute Engine network to which the cluster is connected.
* `operatorNamespace` - Kubernetes namespace where the Akka Cloud Platform operator is running

Cloud SQL Instance

* `postgreSQLEndpoint` - The private IP through which direct connectivity between GKE and Cloud SQL instance is set up
* `postgreSQLInstanceConnectionName` - The Postgresql instance connection name
* `postgreSQLInstanceName` - The Postgresql instance name 

== Pulumi Playbook configuration

You can configure the playbook with the following syntax at the command line:

[source,shell script]
----
pulumi config set <CONFIG KEY> <CONFIG VALUE>`
----

Or by populating the stack's `Pulumi.<STACK NAME>.yaml`.

.Table Configuration
|===
|Configuration Key | Description | Required | Default value

| `gcp:region`
| The AWS region for the Google GKE cluster and other services.

> | No

Reviewer 3: What is the default then?

Author: There is no default here. It can be optionally set through environment variables or `pulumi config`, otherwise it continues with some provider default, which we don't set

A new installation will take approximately  _20 minutes_ and the following components will be provisioned:

> 1. Create a Google GKE cluster.

Reviewer 4: These don't correspond with the steps that follow. To be paralell with the lead-in sentence, they should be regular bullets:

* A Google GKE cluster
* The Akka Cloud Platform Helm chart
* A PostgreSQL Database Instance

Author: 👍

> In addition, in order to operate the cluster, you need to setup:

Reviewer 4: s/be "set up"

Author: 👍

gcloud config set project <YOUR_GCP_PROJECT_ID>
----

> * Pulumi requires that you setup default application credentials in order to interact with your Google Cloud resources. Pulumi requires this.

Reviewer 4: "set up" instead of setup. And the "Pulumi requires this." is unnecessary, it just repeats what is said in the previous paragraph.

Reviewer 2: ```suggestion
*  You are required to set up default application credentials in order to interact with your Google Cloud resources. Pulumi requires this.
```

Author: 👍

[source, shell script]
----
gcloud auth application-default login
----

> * Some optional settings can be set up to provide ready defaults. The following environment variables can be set to provide optional configurations to the Pulumi Google Cloud Platform provider:

Reviewer 4: ```suggestion
* You can set the following environment variables to provide default options to the Pulumi Google Cloud Platform provider:
```

Reviewer 2: I think the suggestion is may be truncated.
```suggestion
* Optionally, you can set the following environment variables to provide preset default configurations for the Pulumi Google Cloud Platform provider:
```

Author: 👍

== Quick Start with Pulumi Playbook

Pulumi is a cloud provisioning tool used to set up cloud infrastructure and Kubernetes cluster resources.

> The fastest way to get started with Akka Cloud Platform on GCP is to run the Akka Cloud Platform Deployment Pulumi playbook, which contains a bunch of instructions that plays like a book when run and on successful completion provisions the cluster and all the appropriate artifacts.

Reviewer 4: ```suggestion
The fastest way to get started with Akka Cloud Platform on GCP is to run the Akka Cloud Platform Deployment Pulumi playbook. The playbook contains a set of instructions that will provision the cluster and all the appropriate artifacts.
```

Reviewer 2: I think that running the playbook has to be explicit.

```suggestion
The fastest way to get started with Akka Cloud Platform on GCP is to run the Akka Cloud Platform Deployment Pulumi playbook. The playbook contains a set of instructions that, when run, will provision the cluster and all the appropriate artifacts.
```

Author: 👍

> Here are the steps to follow to have your playbook up 'n running:

Reviewer 4: and instead of 'n?

Author: 👍

> * Download and install the https://www.pulumi.com/docs/get-started/install/[Pulumi {tab-icon}, window="tab"] cloud engineering (provisioning) tool for your platform.

Reviewer 4: This is different from the playbook? It's a bit confusing.

Author: It's different from the playbook. It's the engine which will run the playbook

> * Download (clone) the https://github.com/lightbend/akka-cloud-platform-deploy[Akka Cloud Platform Deployment Pulumi Playbook {tab-icon}, window="tab"] with https://git-scm.com/book/en/v2/Getting-Started-Installing-Git[`git` {tab-icon}, window="tab"].

Reviewer 4: is (clone) necessary?

Author: Yes, the user needs to _clone_ the git repo to download the playbook

[source,shell script]
----
git clone https://github.com/lightbend/akka-cloud-platform-deploy
----

* Navigate to the base directory of the cloned repository in a terminal.

[source,shell script]
----
cd akka-cloud-platform-deploy
----

> * The same repository contains the scripts for AWS and GCP deployment. Navigate to the `gcp` folder.

Reviewer 4: ```suggestion
* The repository contains scripts for both AWS and GCP deployment. Navigate to the `gcp` folder.
```

Author: 👍

[source,shell script]
----
cd gcp
----

* Install `npm` dependencies:

[source,shell script]
----
npm install
----

* Initialize a https://www.pulumi.com/docs/intro/concepts/stack/[Pulumi Stack {tab-icon}, window="tab"].

> A stack is an instance of the deployed Pulumi playbook and its associated state.

Reviewer 4: This too seems like concepts that s/be introduced before the steps.

And, the definition of stack doesn't jibe with how a playbook is described above. Maybe:

A stack is the cluster and its artifacts with associated state, as provisioned by a Pulumi playbook.

Author: 👍

For example, you might use different stacks to represent different environments such as `dev`, `testing`, and `staging`.
State is managed by default using Pulumi's state management service (a remote service), but you can choose a https://www.pulumi.com/docs/intro/concepts/state/#backends[self-managed state backend {tab-icon}, window="tab"] if you like.
Using Pulumi's service is the fastest way to get started and provides several other benefits such as sharing state with others, view progress of a deployment, and view output variables and the resulting cloud resources that were created.
To learn more about https://www.pulumi.com/docs/intro/concepts/stack/[Pulumi Stack {tab-icon}, window="tab"] and https://www.pulumi.com/docs/intro/concepts/state/[Pulumi State Management {tab-icon}, window="tab"] see the Pulumi documentation.

[source,shell script]
----
pulumi stack init
----

> * Choose Google Project Id, zone and region to use:

Reviewer 4: Choose "the"

Author: 👍

NOTE: The setting of project id and zone are mandatory for a GCP installation. In case these are not set, the installation will be aborted with appropriate error messages.

> * Setup the license key file path. This is required to be set before you run the playbook. This sets up the license key file path which will be passed to the operator deployment process. If you are deploying from local environment you need to download the license file once from GCP Marketplace as per instructions in the xref:gcp-install.adoc[installation guide]. In the {gcp-marketplace}[Akka Cloud Platform {tab-icon}, window="tab"] for GCP page, click on __Configure__ and then go to the tab __Deploy via command line__ and click on __Generate license key__. A `license.yaml` will be generated on your local system.

Reviewer 4: ```suggestion
* Set up the license key file path. This is required before you run the playbook. It will be passed to the operator deployment process. If you are deploying from your local environment you need to download the license file once from GCP Marketplace following the instructions in the xref:gcp-install.adoc[installation guide]. In the {gcp-marketplace}[Akka Cloud Platform {tab-icon}, window="tab"] for GCP page, click on __Configure__ and then go to the tab __Deploy via command line__ and click on __Generate license key__. A `license.yaml` will be generated on your local system.
```

Author: 👍

[source,shell script]
----
pulumi config set akka-cloud-platform-gcp-deploy:license-file-path "/<path>/<to>/license.yaml"
----

> * Setup the version of the operator.

Reviewer 4: Set instead of set up?

Author: 👍

[source,shell script,subs="attributes+"]
----
pulumi config set akka-cloud-platform-gcp-deploy:operator-version {akka-cloud-platform-version}
----

> NOTE: If you forget to set the license key file path and operator version up, the playbook will give you a gentle reminder to set the values appropriately.

Reviewer 4: Is this really necessary? If they follow our steps, they will never see the reminder and if they don't and see the reminder, we don't really need to tell them about it, do we?

Reviewer 2: It is nice to have as an assurance if you are reading through the doc separate from when actually working through the steps.

When you run a `pulumi up` (deploy) all the resources defined in the playbook will be provisioned to your target environment.

> In almost all cases resources will start with a prefix `acp-<STACK NAME>`, where `acp` stands for "Akka Cloud Platform" and `<STACK NAME>` is the name you chose for the Pulumi stack.

Reviewer 4: ```suggestion
In almost all cases, resources will start with a prefix `acp-<STACK NAME>`, where `acp` stands for "Akka Cloud Platform" and `<STACK NAME>` is the name you chose for the Pulumi stack.
```

Author: 👍

  error: update failed
----

> Doing a `pulumi cancel` will cancel the running update. However in some cases the update might have gone __'OK'__ despite the message mentioned above. In that case, `pulumi cancel` gives the following message __The Update has already completed__.

Reviewer 4: ```suggestion
Entering `pulumi cancel` will cancel the running update. However, in some cases, the update might have a status of  'OK', contrary to the message mentioned above. In that case, `pulumi cancel` gives the following message `The Update has already completed`.
```

Author: 👍
